src/StaticAnalysis/ExecutableLinesFindingVisitor.php

```diff
@@ -242,6 +247,29 @@ private function getLines(Node $node): array
         return [$node->getStartLine()];
     }
 
+    private function getNodeStartLine(NodeAbstract $node): int
+    {
+        if ($node instanceof Node\Expr\BooleanNot ||
```

Comment:
what have these operators in common? can you suggest a short comment to place above the if, so the code can be more understandable?

https://www.php.net/manual/en/language.operators.precedence.php

https://github.com/nikic/PHP-Parser/tree/v4.15.1/lib/PhpParser/Node/Expr

Reply:
I have no idea. I put all my attention in crafting the tests and making them as close to reality as possible.
Then I blindly adjust the `src/` to get them passing :wink:
Of course I tried to refactor the code looking for repeating patterns at exactly the two links you've also followed, but I found none :frowning_face:

Reply:
I hunted more test cases and the next line with `Node\Expr\AssignOp\ShiftLeft` is not even needed.

The `Node\Expr\BooleanNot` condition is fixing one test, but breaking other I newly added (`complexAssociativityNa2`). I fixed it by calling `getNodeStartLine` method instead of `$node->getEndLine()`, but other test case `unaryMinusNowdoc` breaks tests again.

Currently the static analyser heavily relies on scalars beiing single line. But this is an incorrect assumption and position of all scalars must be analysed using `getNodeStartLine` to handle heredoc/nowdoc correctly.

Then no special handling for `Node\Expr\BooleanNot` should be needed, more precisely, all unary operators has to be enumerated, as there is no `UnaryOp` interface - https://github.com/nikic/PHP-Parser/blob/v4.15.1/lib/PhpParser/Node/Expr/BooleanNot.php#L7.

Reply:
> I hunted more test cases

I'd not going down the rabbit hole too much: just stick to real cases, we can't handle every edge cases.
Being 99.9% effective on real existing code is much easier and simplier than aimint at 100% correctness with a red build

> also `testHeavyIndentationIsHandledCorrectly` test seems to output lines for assign that are not present in xdebug output

That one was my fist test coming from a real example but expanded by my imagination: it was a mistake becase it's not easy to understand and doesn't represent the reality; I've kept it though because it's still a good regression test.


I suggest to keep the good https://github.com/sebastianbergmann/php-code-coverage/pull/949/commits/9fc658244bd7c049fc65785dcc5de61780c79ec5 but to revert https://github.com/sebastianbergmann/php-code-coverage/pull/949/commits/13bc95351a7ae890b19e04fafc67f17fd698d8e0 and let's go live with the fixes with managed to accomplish

Reply:
> I'd not going down the rabbit hole too much: just stick to real cases, we can't handle every edge cases

I could not agree more. Scalars split across multiple lines are definitely not something we should care about.

Reply:
simple test case as in `testHeavyIndentationIsHandledCorrectly` like

```
    public function xx(): void
    {
        $v3
            =
            x()
        ;
    }
```

must not emit executable line for assign as no such line is emit with xdebug

I will look into this problem by myself and ask for help if needed.

Reply:
Nobody should write code like that. And if they do, they're on their own. We should not care about this.

Reply:
@Slamdunk please review

when resolving any php parser node to line, `getNodeStartLine` method must be always used